File: internal/api/jdwp/classobjectreference.go
package jdwp

import (
	"context"
	"fmt"

	"cli-debugger/internal/api"
)

// ClassObjectReference Command Set Implementation
// ClassObjectReference command constants are already defined in protocol.go

// ReflectedType Get reflected type
func (c *Client) ReflectedType(classObjectID string) (string, error) {
	data := encodeID(classObjectID, c.idsizes.ObjectIDSize)

	packet := createCommandPacketWithData(classObjectCommandSet, classObjectCommandReflectedType, data)
	if err := c.sendPacket(packet); err != nil {
		return "", &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get reflected type",
			Cause:   err,
		}
	}

	reply, err := c.readReply()
	if err != nil {
		return "", &api.APIError{
			Type:    api.CommandError,
			Message: "Failed to get reflected type",
			Cause:   err,
		}
	}

	if reply.ErrorCode != 0 {
		return "", &api.APIError{
			Type:    api.ProtocolError,
			Code:    int(reply.ErrorCode),
			Message: fmt.Sprintf("Get reflected type failed: %s", reply.Message),
		}
	}

	reader := newPacketReader(reply.Data)
	tag := reader.readByte()
	refTypeID := reader.readID(c.idsizes.ReferenceTypeIDSize)

	return fmt.Sprintf("%c:%s", tag, refTypeID), nil
}

File: src/protocol/jdwp/class-object-reference.ts
/**
 * ClassObjectReference Command Set Implementation
 */

import {
  CommandSet,
  ClassObjectReferenceCommand,
  type IDSizes,
} from "./protocol.js";
import {
  createCommandPacketWithData,
  encodeID,
} from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get reflected type for a class object
 */
export async function getReflectedType(
  executor: JDWPCommandExecutor,
  classObjectID: string,
): Promise<string> {
  const data = encodeID(classObjectID, executor.idSizes.objectIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ClassObjectReference,
    ClassObjectReferenceCommand.ReflectedType,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get reflected type failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readID(executor.idSizes.referenceTypeIDSize);
}
